File: server/event-filters.ts
/**
 * Event filtering utilities for EPCAL calendar rendering.
 *
 * Provides functions to filter and organize events for specific days.
 */

import { startOfDay, isSameDay } from "date-fns";

/**
 * Calendar event interface
 */
export interface CalendarEvent {
  id: string;
  title: string;
  start: Date;
  end: Date;
  allDay?: boolean;
  calendarColor?: string;
  calendarIcon?: string; // Icon/letter to identify the calendar
  calendarId?: string; // Entity ID of the calendar this event belongs to
}

/**
 * Event with day-specific indicators
 */
export interface EventForDay {
  event: CalendarEvent;
  startsOnDay: boolean;
  endsOnDay: boolean;
}

/**
 * Collection type configuration
 */
export interface CollectionType {
  name: string;
  icon: string;
}

/**
 * Filters events for a specific day and returns them with indicators
 * showing whether each event starts/ends on that day.
 *
 * Includes:
 * - Events that start on this day
 * - Multi-day events that span across this day (started before, end on/after)
 *
 * @param events - Array of calendar events
 * @param day - Date to filter for
 * @returns Events for the day with start/end indicators
 */
export function getEventsForDay(events: CalendarEvent[], day: Date): EventForDay[] {
  const dayStart = startOfDay(day);

  return events
    .filter((e) => {
      const startsOnDay = isSameDay(e.start, day);
      const spansDay = e.start < dayStart && e.end >= dayStart;
      return startsOnDay || spansDay;
    })
    .map((e) => ({
      event: e,
      startsOnDay: isSameDay(e.start, day),
      endsOnDay: isSameDay(e.end, day),
    }));
}

/**
 * Gets collection calendar icons for a specific day
 * Returns icons for each collection type found in events on this day
 *
 * @param events - Array of calendar events
 * @param day - Date to check for collection events
 * @param collectionCalendars - Array of calendar IDs that are collection calendars
 * @param collectionTypes - Array of collection type configurations with icons
 * @returns Array of icon names for collections on this day
 */
export function getCollectionIconsForDay(
  events: CalendarEvent[],
  day: Date,
  collectionCalendars: string[] = [],
  collectionTypes: CollectionType[] = [],
): string[] {
  if (collectionCalendars.length === 0 || collectionTypes.length === 0)
    return [];

  const icons: string[] = [];
  const foundTypes = new Set<string>();

  // Find all collection events on this day
  const collectionEvents = events.filter(
    (e) =>
      e.calendarId &&
      collectionCalendars.includes(e.calendarId) &&
      e.allDay &&
      isSameDay(e.start, day),
  );

  // Keywords for matching event titles to collection types
  const typeKeywords: { [typeName: string]: string[] } = {
    Garbage: ["garbage", "trash", "waste", "déchet", "ordure", "poubelle"],
    Recycling: ["recycling", "recycle", "recyclage", "récupération"],
    Compost: ["compost", "organic", "organique", "food waste"],
    "Yard Waste": ["yard", "green", "vert", "garden"],
    Glass: ["glass", "verre"],
    Paper: ["paper", "papier"],
    Cardboard: ["cardboard", "carton"],
  };

  // Match each event title against collection types using keywords
  for (const event of collectionEvents) {
    const titleLower = event.title.toLowerCase();

    for (const type of collectionTypes) {
      const keywords = typeKeywords[type.name] || [type.name.toLowerCase()];

      if (
        keywords.some((kw) => titleLower.includes(kw)) &&
        !foundTypes.has(type.name)
      ) {
        icons.push(type.icon);
        foundTypes.add(type.name);
      }
    }
  }

  return icons;
}

File: custom_components/epcal/renderer/event_filters.py
"""Event filtering utilities for EPCAL calendar rendering.

Provides functions to filter and organize events for specific days.
"""

from datetime import datetime, time
from typing import TypedDict

from .types import CalendarEvent


class EventForDay(TypedDict):
    """Event with day-specific indicators."""

    event: CalendarEvent
    startsOnDay: bool
    endsOnDay: bool


def get_events_for_day(events: list[CalendarEvent], day: datetime) -> list[EventForDay]:
    """Filter events for a specific day with start/end indicators.

    Includes:
    - Events that start on this day
    - Multi-day events that span across this day (started before, end on/after)

    Args:
        events: Array of calendar events (with parsed datetime objects)
        day: Date to filter for

    Returns:
        Events for the day with start/end indicators
    """
    # Ensure day_start has the same timezone awareness as the events
    day_start = datetime.combine(day.date(), time.min)

    # If day is naive but events might be aware, use first event's timezone
    if day.tzinfo is None and events:
        # Check if events have timezone info
        for event in events:
            start = event.get("start")
            if start and hasattr(start, "tzinfo") and start.tzinfo is not None:
                day_start = day_start.replace(tzinfo=start.tzinfo)
                break
    elif day.tzinfo is not None:
        day_start = day_start.replace(tzinfo=day.tzinfo)

    result = []
    for event in events:
        event_start = event.get("start")
        event_end = event.get("end")

        # Skip events without start/end times
        if not event_start or not event_end:
            continue

        # Normalize timezone awareness for comparison
        # If one is aware and the other is naive, make both naive for comparison
        compare_start = event_start
        compare_end = event_end
        compare_day_start = day_start

        # Convert all to naive if there's a mismatch
        if (event_start.tzinfo is None) != (day_start.tzinfo is None):
            compare_start = (
                event_start.replace(tzinfo=None) if event_start.tzinfo else event_start
            )
            compare_end = (
                event_end.replace(tzinfo=None) if event_end.tzinfo else event_end
            )
            compare_day_start = (
                day_start.replace(tzinfo=None) if day_start.tzinfo else day_start
            )

        # Check if event starts on this day
        starts_on_day = event_start.date() == day.date()

        # Check if event spans this day (started before, ends on/after)
        spans_day = (
            compare_start < compare_day_start and compare_end >= compare_day_start
        )

        if starts_on_day or spans_day:
            result.append(
                {
                    "event": event,
                    "startsOnDay": event_start.date() == day.date(),
                    "endsOnDay": event_end.date() == day.date(),
                }
            )

    return result


def get_collection_icons_for_day(
    events: list[CalendarEvent], day: datetime, collection_calendar_ids: list[str]
) -> list[str]:
    """Get collection calendar icons for a specific day.

    Returns icons from collection calendar entities for events on this day.
    Uses the calendar entity's icon directly (no keyword matching).

    Args:
        events: Array of calendar events
        day: Date to check for collection events
        collection_calendar_ids: Array of calendar IDs that are waste collection calendars

    Returns:
        Array of icons for collections on this day (deduplicated)
    """
    if not collection_calendar_ids:
        return []

    icons = []
    seen_icons = set()

    # Find all collection events on this day
    for event in events:
        calendar_id = event.get("calendarId")
        if not calendar_id or calendar_id not in collection_calendar_ids:
            continue

        # Check if event is on this day (all-day events only)
        if not event.get("allDay"):
            continue

        start = event.get("start")
        if not start or start.date() != day.date():
            continue

        # Use the calendar's icon directly
        icon = event.get("calendarIcon")
        if icon and icon not in seen_icons:
            icons.append(icon)
            seen_icons.add(icon)

    return icons
